solidity/contracts/bridge/Bridge.sol

```diff
@@ -1460,68 +1474,38 @@ contract Bridge is Ownable {
-    //            outputs with same script hash. This will produce the same
-    //            redemption key which must be put to `invalidRedemptions`
-    //            multiple times so it should not cause a revert and it is
-    //            enough to count it once in `invalidRedemptions`.
-    //       6. If `invalidCount > 0` transaction is fraudulent. If not - revert.
-    //       7. Punish the wallet, probably by slashing its operators.
-    //       8. Change wallet's state in `wallets` mapping to `MovingFunds` in
+    //       3. Check if wallet state is Active or MovingFunds
+    //       4. Validate the number of inputs. If there is one input and it
+    //          points to the wallet's main UTXO - move to point 5. If there
+    //          are multiple inputs and there is wallet's main UTXO in the set,
+    //          check if this is  a sweep transaction. If it's not a sweep,
+    //          consider it as fraudulent and move to point 6.
+    //          In all other cases revert the call.
+    //       5. Extract the output and analyze its type. The output is not
+    //          a fraud and the call should be reverted ONLY IF one of the
+    //          following conditions is true:
+    //          - Output is a requested redemption held by `pendingRedemptions`
+    //            and output value fulfills the request range. There is an
+    //            open question if a misfunded request should be removed
+    //            from `pendingRedemptions` (probably yes) and whether the
+    //            redeemer should be reimbursed in case of an underfund.
+    //          - Output is a timed out redemption held by `timedOutRedemptions`
+    //            and output value fulfills the request range.
+    //          - Output is a proper change i.e. a single output targeting
+    //            the wallet PKH back and having a non-zero value.
+    //          - Wallet is in MovingFunds state, the output points to the
+    //            expected target wallet, have non-zero value, and is a single
+    //            output in the vector.
+    //          In all other cases consider the transaction as fraud and
+    //          proceed to point 6.
+    //       6. Punish the wallet, probably by severely slashing its operators.
+    //       7. Change wallet's state in `wallets` mapping to `Terminated` in
     //          order to prevent against new redemption requests hitting
-    //          that wallet.
-    //       9. Expect the wallet to transfer its funds to another healthy
-    //          wallet (just as in case of failed heartbeat).
-    //      10. Put the transaction hash to `fraudulentTransactions` mapping to
-    //          prevent against replays.
+    //          that wallet. This also prevents against reporting a fraud
+    //          multiple times for one transaction (see point 3) and blocks
+    //          submission of sweep and redemption proofs.
```

Comment:
We could add it to the end, to ensure it's in the documentation of the function we are going to implement:

`Terminated` wallet is blocked in the Bridge forever. If the fraud was a mistake done by the wallet and the wallet is still honest deep in its heart, the wallet can coordinate off-chain to recover the BTC and donate it to another wallet. If they recover all of the remaining BTC, DAO might decide to reward them with tokens so that they can have at least some portion of their slashed tokens back.

Reply:
See https://github.com/keep-network/tbtc-v2/pull/128/commits/64395ffafef55a2795e927a1dad3c531d0ef24d8